File: learn/stack-and-queue/queue-array/queue_array.cpp
class QueueArray {
private:
    int* arr;
    int front, rear, capacity, size;

public:
    QueueArray(int cap) {
        capacity = cap;
        arr = new int[capacity];
        front = 0;
        rear = -1;
        size = 0;
    }

    ~QueueArray() {
        delete[] arr;
    }

    bool isEmpty() {
        return size == 0;
    }

    bool isFull() {
        return size == capacity;
    }

    void enqueue(int value) {
        if (isFull()) return;
        rear = (rear + 1) % capacity;
        arr[rear] = value;
        size++;
    }

    int dequeue() {
        if (isEmpty()) return -1;
        int value = arr[front];
        front = (front + 1) % capacity;
        size--;
        return value;
    }

    int peek() {
        if (isEmpty()) return -1;
        return arr[front];
    }
};

File: learn/stack-and-queue/queue-array/queue_array.go


type QueueArray struct {
	arr      []int
	front    int
	rear     int
	capacity int
	size     int
}

func NewQueueArray(cap int) *QueueArray {
	return &QueueArray{
		arr:      make([]int, cap),
		capacity: cap,
		front:    0,
		rear:     -1,
		size:     0,
	}
}

func (q *QueueArray) IsEmpty() bool {
	return q.size == 0
}

func (q *QueueArray) IsFull() bool {
	return q.size == q.capacity
}

func (q *QueueArray) Enqueue(value int) {
	if q.IsFull() {
		return
	}
	q.rear = (q.rear + 1) % q.capacity
	q.arr[q.rear] = value
	q.size++
}

func (q *QueueArray) Dequeue() int {
	if q.IsEmpty() {
		return -1
	}
	value := q.arr[q.front]
	q.front = (q.front + 1) % q.capacity
	q.size--
	return value
}

func (q *QueueArray) Peek() int {
	if q.IsEmpty() {
		return -1
	}
	return q.arr[q.front]
}
